File: web/src/model/placeLocation.ts
/**
 * A geographic point, optionally named.
 *
 * Ported 1:1 from `model/PlaceLocation.kt`.
 */

import { type SavedPlace, savedPlaceFromCoordinates } from "./savedPlace";

export interface PlaceLocation {
  readonly latitude: number;
  readonly longitude: number;
  readonly name?: string;
}

export function toRouteValue(location: PlaceLocation): string {
  return `${location.latitude.toFixed(6)},${location.longitude.toFixed(6)}`;
}

export function toSavedPlace(location: PlaceLocation): SavedPlace {
  const coordinatePlace = savedPlaceFromCoordinates(location.latitude, location.longitude);
  const routeName = location.name?.trim();
  if (routeName !== undefined && routeName.length > 0) {
    return { ...coordinatePlace, name: routeName };
  }
  return coordinatePlace;
}

export function placeLocationFromSavedPlace(place: SavedPlace): PlaceLocation {
  return { latitude: place.latitude, longitude: place.longitude };
}

export function placeLocationFromRouteValue(value: string): PlaceLocation | null {
  const parts = value.split(",");
  if (parts.length !== 2) return null;

  const latitude = parseCoordinate(parts[0]);
  const longitude = parseCoordinate(parts[1]);
  if (latitude === null || longitude === null) return null;
  if (!Number.isFinite(latitude) || latitude < -90.0 || latitude > 90.0) return null;
  if (!Number.isFinite(longitude) || longitude < -180.0 || longitude > 180.0) return null;

  return { latitude, longitude };
}

/**
 * Mirrors Kotlin `String.toDoubleOrNull()`: returns null for blank or
 * non-numeric input instead of JavaScript's `NaN`.
 */
function parseCoordinate(raw: string | undefined): number | null {
  if (raw === undefined) return null;
  const trimmed = raw.trim();
  if (trimmed.length === 0) return null;
  const value = Number(trimmed);
  return Number.isNaN(value) ? null : value;
}

File: shared/src/commonMain/kotlin/com/cloudbasepredictor/model/PlaceLocation.kt
package com.cloudbasepredictor.model

import com.cloudbasepredictor.util.toFixedDecimalString

data class PlaceLocation(
    val latitude: Double,
    val longitude: Double,
    val name: String? = null,
) {
    fun toRouteValue(): String {
        return "${latitude.toFixedDecimalString(ROUTE_DECIMAL_PLACES)}," +
            longitude.toFixedDecimalString(ROUTE_DECIMAL_PLACES)
    }

    fun toSavedPlace(): SavedPlace {
        val coordinatePlace = SavedPlace.fromCoordinates(
            latitude = latitude,
            longitude = longitude,
        )
        val routeName = name?.trim()?.takeIf { it.isNotEmpty() }
        return if (routeName != null) {
            coordinatePlace.copy(name = routeName)
        } else {
            coordinatePlace
        }
    }

    companion object {
        private const val ROUTE_DECIMAL_PLACES = 6

        fun fromSavedPlace(place: SavedPlace): PlaceLocation {
            return PlaceLocation(
                latitude = place.latitude,
                longitude = place.longitude,
            )
        }

        fun fromRouteValue(value: String): PlaceLocation? {
            val parts = value.split(',')
            if (parts.size != 2) return null

            val latitude = parts[0].toDoubleOrNull() ?: return null
            val longitude = parts[1].toDoubleOrNull() ?: return null
            if (!latitude.isFinite() || latitude !in -90.0..90.0) return null
            if (!longitude.isFinite() || longitude !in -180.0..180.0) return null

            return PlaceLocation(latitude = latitude, longitude = longitude)
        }
    }
}
